test/jit/test_hooks.py

```diff
@@ -0,0 +1,757 @@
+            return output
+
+        m.submodule.register_forward_pre_hook(pre_hook1)
+        m.submodule.register_forward_pre_hook(pre_hook2)
+        m.submodule.register_forward_hook(forward_hook1)
+        m.submodule.register_forward_hook(forward_hook2)
+
+        self.checkModule(m, (["a"]))
+
+    def test_nested_tuple_IO(self):
+        m = OuterModuleTupleDoubleIO("outer_mod_name", "inner_mod_name")
+
+        def pre_hook_outermod(self, input: Tuple[Tuple[str, str]]):
+            # 'return ("hello", "goodbye")' doesn't work with eager because
+            # tuple is unpacked by eager when forward isn't expecting it
+            return (("hello", "goodbye"),)
+
+        def pre_hook_innermod(self, input: Tuple[Tuple[str, str]]):
+            # 'return ("hey","howdy")' doesn't work with eager because
+            # tuple unpacked by eager when forward isn't expecting it
+            return (("hey", "howdy"),)
+
+        def forward_hook_outermod(self, input: Tuple[Tuple[str, str]], output: str):
+            return ("a", "b")
+
+        def forward_hook_innermod(
+            self, input: Tuple[Tuple[str, str]], output: Tuple[str, str]
+        ):
+            return "forward_inner_mod"
+
+        m.register_forward_pre_hook(pre_hook_outermod)
+        m.submodule.register_forward_pre_hook(pre_hook_innermod)
+        m.register_forward_hook(forward_hook_outermod)
+        m.submodule.register_forward_hook(forward_hook_innermod)
+
+        self.checkModule(m, (("a", "b"),))
+
+        m = OuterModuleTupleSingleIO("outer_mod_name", "inner_mod_name")
+
+        def pre_hook_outermod(self, input: Tuple[Tuple[int]]):
+            # 'return (11,)' doesn't work with eager, inner tuple lost
+            return ((11,),)
+
+        def pre_hook_innermod(self, input: Tuple[Tuple[int]]):
+            # 'return (22,)' doesn't work with eager, inner tuple lost
+            return ((22,),)
+
+        def forward_hook_outermod(self, input: Tuple[Tuple[int]], output: int):
+            return (11,)
+
+        def forward_hook_innermod(self, input: Tuple[Tuple[int]], output: Tuple[int]):
+            return 22
+
+        m.register_forward_pre_hook(pre_hook_outermod)
+        m.submodule.register_forward_pre_hook(pre_hook_innermod)
+        m.register_forward_hook(forward_hook_outermod)
+        m.submodule.register_forward_hook(forward_hook_innermod)
+
+        self.checkModule(m, ((3,),))
+
+    def test_hook_method_name_collision(self):
+        m = OuterModuleSingleIO("outer_mod_name", "inner_mod_name")
+
+        def foo(self, input: Tuple[str]):
+            assert self.name == "inner_mod_name"
+            assert input[0] == "a_outermod"
+            return ("pre_hook_overrid_name",)
+
+        m.submodule.register_forward_pre_hook(foo)
+
+        err_msg = (
+            "Can't define hook: foo on class: .+ "
+            "because a method or hook with that name already exists."
```

Comment:
So this tests collisions between hooks and methods, but what about hooks and other hooks? I guess redefining a hook function is tricky, but something like this might work:
```
def hook_fn(...):
   ...

mod.register_forward_pre_hook(hook_fn)

def hook_fn(...):
   ...
```
I know the interpreter accepts a redefinition like this, but I'm not sure about redefinitions in source.

Reply:
Good catch, the existing code allowed this and was running the first hook definition twice. I changed the logic in _recursive.py 'get_hook_stubs' on the bottom commit to check for this. I also added the test case to check for it.